### PR TITLE
Use RawBytes and co in RawPage

### PR DESCRIPTION
**Pull request by jorisdral**

### CODE REVIEW

File: src/Database/LSMTree/Internal/Serialise/RawBytes.hs [hunk 1]
```diff
@@ -146,6 +148,12 @@ sizeofRawBytes (RawBytes pvec) = P.length pvec
 concat :: [RawBytes] -> RawBytes
 concat rbs = RawBytes (P.concat (fmap toVector rbs))
 
+fromList :: [Word8] -> RawBytes
+fromList = coerce P.fromList
+
+toList :: RawBytes -> [Word8]
```
Comment (phadej) [inline src/Database/LSMTree/Internal/Serialise/RawBytes.hs:154]: I'd name these `pack` and `unpack` (as in `Text` and `ByteString`). `toList` clashes with `Foldable` method, it's a bad name when names are unqualified.

(Alternatively or additionally, `RawBytes` could have instance of `GHC.Exts.IsList`)

File: test/Test/Database/LSMTree/Internal/RawPage.hs [hunk 2]
```diff
@@ -293,8 +297,8 @@ prop_single_entry k op blobref = label (show $ BS.null sfx) $
     sfxSize :: Int
     sfxSize = BS.length sfx
 
-bsToVector :: BS.ByteString -> P.Vector Word8
-bsToVector = P.fromList . BS.unpack
+bsToRawBytes :: BS.ByteString -> RawBytes
```
Comment (phadej) [inline test/Test/Database/LSMTree/Internal/RawPage.hs:300]: There is function for this and `bsFromRawBytes`, in `RawBytes` module, isn't there?

If there aren't, there probably should be.

File: src/Database/LSMTree/Internal/RawPage.hs [hunk 3]
```diff
@@ -62,9 +66,9 @@ makeRawPage ba off = RawPage (div2 off) ba
 
 rawPageEntry
     :: RawPage
-    -> P.Vector Word8 -- ^ key
-    -> Maybe (Entry (P.Vector Word8) BlobSpan)
-rawPageEntry !page !key = bisect 0 (fromIntegral dirNumKeys)
+    -> SerialisedKey
+    -> Maybe (Entry SerialisedValue BlobSpan)
+rawPageEntry !page (SerialisedKey' !key) = coerce $ bisect 0 (fromIntegral dirNumKeys)
```
Comment (dcoutts) [inline src/Database/LSMTree/Internal/RawPage.hs:71]: Can we use proper constructors please, not `coerce`.

**Review by dcoutts — COMMENTED** [review]
So I had been planning on doing something like this, but using `RawBytes` (or even `SerialisedKey`/`SerialisedValue` everywhere within the RawPage code, not just on the boundary.

In particular, I suspect (but I've not benchmarked) that the `P.Vector Word8` equality and comparison operators are not as good as for `RawBytes` which overrides the `Eq` and `Ord` instance to use `compareBytes`. The `P.Vector` instances use streaming code, not FFI to `memcmp`.

File: src/Database/LSMTree/Internal/RawPage.hs (hunk 3 above)
Comment (phadej) [reply src/Database/LSMTree/Internal/RawPage.hs:71]: In general we cannot, this coerces from `Maybe (Entry (P.Vector Word8) BlobSpan` to `Maybe (Entry SerialisedValue BlobSpan)`.

But in this specific case, sure, we won't need to `coerce` if `bisect`, `found` etc returned types with `SerialisedValue`

Comment (jorisdral) [reply src/Database/LSMTree/Internal/RawPage.hs:71]: I didn't want to cause too much churn in this PR, but I'm happy to change everything to `Serialised*` constructors

**Comment by jorisdral** [comment]
> So I had been planning on doing something like this, but using `RawBytes` (or even `SerialisedKey`/`SerialisedValue` everywhere within the RawPage code, not just on the boundary.
> 
> In particular, I suspect (but I've not benchmarked) that the `P.Vector Word8` equality and comparison operators are not as good as for `RawBytes` which overrides the `Eq` and `Ord` instance to use `compareBytes`. The `P.Vector` instances use streaming code, not FFI to `memcmp`.

It seems indeed that the intra-page lookups become faster when switching to `RawBytes` and co

<details>
<summary> Before </summary>

```
❯ cabal run lsm-tree-micro-bench -- -m pattern "RawPage" --regress allocated:iters +RTS -T -RTS
benchmarking Bench.Database.LSMTree.Internal.RawPage/missing
time                 37.51 ns   (37.30 ns .. 37.82 ns)
                     0.999 R²   (0.999 R² .. 1.000 R²)
mean                 37.96 ns   (37.63 ns .. 38.56 ns)
std dev              1.419 ns   (990.9 ps .. 2.126 ns)
allocated:           1.000 R²   (1.000 R² .. 1.000 R²)
  iters              112.000    (112.000 .. 112.000)
  y                  3084.683   (2885.082 .. 3284.189)
variance introduced by outliers: 59% (severely inflated)

benchmarking Bench.Database.LSMTree.Internal.RawPage/existing-head
time                 77.48 ns   (77.35 ns .. 77.64 ns)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 77.36 ns   (77.27 ns .. 77.52 ns)
std dev              405.2 ps   (278.1 ps .. 638.5 ps)
allocated:           1.000 R²   (1.000 R² .. 1.000 R²)
  iters              576.000    (576.000 .. 576.000)
  y                  3080.741   (2872.892 .. 3302.618)

benchmarking Bench.Database.LSMTree.Internal.RawPage/existing-last
time                 84.02 ns   (83.63 ns .. 84.51 ns)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 83.43 ns   (83.27 ns .. 83.69 ns)
std dev              704.2 ps   (499.9 ps .. 1.126 ns)
allocated:           1.000 R²   (1.000 R² .. 1.000 R²)
  iters              576.000    (576.000 .. 576.000)
  y                  3056.873   (2841.658 .. 3275.663)
```
</details>

<details>
<summary> After </summary>

```
❯ cabal run lsm-tree-micro-bench -- -m pattern "RawPage" --regress allocated:iters +RTS -T -RTS
benchmarking Bench.Database.LSMTree.Internal.RawPage/missing
time                 49.81 ns   (49.54 ns .. 50.03 ns)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 49.72 ns   (49.58 ns .. 49.89 ns)
std dev              537.8 ps   (468.1 ps .. 619.5 ps)
allocated:           1.000 R²   (1.000 R² .. 1.000 R²)
  iters              80.000     (80.000 .. 80.000)
  y                  3083.323   (2887.312 .. 3290.036)
variance introduced by outliers: 10% (moderately inflated)

benchmarking Bench.Database.LSMTree.Internal.RawPage/existing-head
time                 53.94 ns   (53.81 ns .. 54.13 ns)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 54.16 ns   (53.97 ns .. 54.46 ns)
std dev              771.7 ps   (546.1 ps .. 1.160 ns)
allocated:           1.000 R²   (1.000 R² .. 1.000 R²)
  iters              176.000    (176.000 .. 176.000)
  y                  3066.674   (2862.763 .. 3287.122)
variance introduced by outliers: 16% (moderately inflated)

benchmarking Bench.Database.LSMTree.Internal.RawPage/existing-last
time                 61.48 ns   (61.29 ns .. 61.66 ns)
                     1.000 R²   (1.000 R² .. 1.000 R²)
mean                 61.59 ns   (61.36 ns .. 61.93 ns)
std dev              927.0 ps   (679.3 ps .. 1.244 ns)
allocated:           1.000 R²   (1.000 R² .. 1.000 R²)
  iters              176.000    (176.000 .. 176.000)
  y                  3069.509   (2868.193 .. 3289.438)
variance introduced by outliers: 18% (moderately inflated)
```
</details>

**Comment by dcoutts** [comment]
> It seems indeed that the intra-page lookups for existing keys become faster when switching to RawBytes and co, but for missing keys it becomes slightly slower. Allocated bytes go down across the board

That's a nice surprise!

**Review by dcoutts — APPROVED** [review]

**Pull request merged**